File: client-sdks/sockudo-dotnet/src/Sockudo.Client/FossilDelta.cs
namespace Sockudo.Client;

public static class FossilDelta
{
    private const int HashWidth = 16;
    private const string Digits = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ_abcdefghijklmnopqrstuvwxyz~";
    private static readonly Dictionary<byte, int> Values = Digits
        .Select((value, index) => new KeyValuePair<byte, int>((byte)value, index))
        .ToDictionary(entry => entry.Key, entry => entry.Value);

    private sealed class Reader
    {
        private readonly byte[] _data;
        public int Position { get; set; }

        public Reader(byte[] data)
        {
            _data = data;
        }

        public bool HasBytes => Position < _data.Length;

        public byte ReadByte()
        {
            if (Position >= _data.Length)
            {
                throw new DeltaFailure("out of bounds");
            }
            return _data[Position++];
        }

        public char ReadCharacter() => (char)ReadByte();

        public int ReadInteger()
        {
            var value = 0;
            while (HasBytes)
            {
                var raw = ReadByte();
                if (!Values.TryGetValue(raw, out var mapped))
                {
                    Position -= 1;
                    break;
                }
                value = (value << 6) + mapped;
            }
            return value;
        }
    }

    public static byte[] Apply(byte[] @base, byte[] delta)
    {
        var reader = new Reader(delta);
        var outputSize = reader.ReadInteger();
        if (reader.ReadCharacter() != '\n')
        {
            throw new DeltaFailure("size integer not terminated by newline");
        }

        var output = new List<byte>(outputSize);
        var total = 0;
        while (reader.HasBytes)
        {
            var count = reader.ReadInteger();
            switch (reader.ReadCharacter())
            {
                case '@':
                    {
                        var offset = reader.ReadInteger();
                        if (reader.HasBytes && reader.ReadCharacter() != ',')
                        {
                            throw new DeltaFailure("copy command not terminated by comma");
                        }
                        total += count;
                        if (total > outputSize)
                        {
                            throw new DeltaFailure("copy exceeds output file size");
                        }
                        if (offset + count > @base.Length)
                        {
                            throw new DeltaFailure("copy extends past end of input");
                        }
                        output.AddRange(@base.Skip(offset).Take(count));
                        break;
                    }
                case ':':
                    {
                        total += count;
                        if (total > outputSize)
                        {
                            throw new DeltaFailure("insert command gives an output larger than predicted");
                        }
                        if (reader.Position + count > delta.Length)
                        {
                            throw new DeltaFailure("insert count exceeds size of delta");
                        }
                        output.AddRange(delta.Skip(reader.Position).Take(count));
                        reader.Position += count;
                        break;
                    }
                case ';':
                    {
                        var bytes = output.ToArray();
                        if (count != Checksum(bytes))
                        {
                            throw new DeltaFailure("bad checksum");
                        }
                        if (total != outputSize)
                        {
                            throw new DeltaFailure("generated size does not match predicted size");
                        }
                        return bytes;
                    }
                default:
                    throw new DeltaFailure("unknown delta operator");
            }
        }
        throw new DeltaFailure("unterminated delta");
    }

    private static int Checksum(byte[] data)
    {
        var sum0 = 0;
        var sum1 = 0;
        var sum2 = 0;
        var sum3 = 0;
        var index = 0;
        var remaining = data.Length;

        while (remaining >= HashWidth)
        {
            sum0 += data[index + 0] + data[index + 4] + data[index + 8] + data[index + 12];
            sum1 += data[index + 1] + data[index + 5] + data[index + 9] + data[index + 13];
            sum2 += data[index + 2] + data[index + 6] + data[index + 10] + data[index + 14];
            sum3 += data[index + 3] + data[index + 7] + data[index + 11] + data[index + 15];
            index += HashWidth;
            remaining -= HashWidth;
        }

        while (remaining >= 4)
        {
            sum0 += data[index + 0];
            sum1 += data[index + 1];
            sum2 += data[index + 2];
            sum3 += data[index + 3];
            index += 4;
            remaining -= 4;
        }

        sum3 += (sum2 << 8) + (sum1 << 16) + (sum0 << 24);
        switch (remaining)
        {
            case 3:
                sum3 += data[index + 2] << 8;
                sum3 += data[index + 1] << 16;
                sum3 += data[index + 0] << 24;
                break;
            case 2:
                sum3 += data[index + 1] << 16;
                sum3 += data[index + 0] << 24;
                break;
            case 1:
                sum3 += data[index + 0] << 24;
                break;
        }

        return sum3;
    }
}

public sealed class DeltaFailure : SockudoException
{
    public DeltaFailure(string message) : base(message)
    {
    }
}

File: client-sdks/sockudo-kotlin/lib/src/main/kotlin/io/sockudo/client/FossilDelta.kt
package io.sockudo.client

object FossilDelta {
    private const val nHash = 16
    private val digits = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ_abcdefghijklmnopqrstuvwxyz~".map { it.code.toByte() }
    private val values =
        IntArray(128) { -1 }.apply {
            digits.forEachIndexed { index, value ->
                this[value.toInt()] = index
            }
        }

    private class Reader(
        private val data: ByteArray,
    ) {
        var position: Int = 0
        val hasBytes: Boolean
            get() = position < data.size

        fun byte(): Byte {
            if (position >= data.size) {
                throw SockudoException.DeltaFailure("out of bounds")
            }
            return data[position++]
        }

        fun character(): Char = byte().toInt().toChar()

        fun integer(): Int {
            var value = 0
            while (hasBytes) {
                val raw = byte().toInt() and 0xff
                val mapped = if (raw < values.size) values[raw] else -1
                if (mapped < 0) {
                    position -= 1
                    break
                }
                value = (value shl 6) + mapped
            }
            return value
        }
    }

    fun apply(base: ByteArray, delta: ByteArray): ByteArray {
        val reader = Reader(delta)
        val outputSize = reader.integer()
        if (reader.character() != '\n') {
            throw SockudoException.DeltaFailure("size integer not terminated by newline")
        }

        val output = ArrayList<Byte>(outputSize)
        var total = 0

        while (reader.hasBytes) {
            val count = reader.integer()
            when (reader.character()) {
                '@' -> {
                    val offset = reader.integer()
                    if (reader.hasBytes && reader.character() != ',') {
                        throw SockudoException.DeltaFailure("copy command not terminated by comma")
                    }
                    total += count
                    if (total > outputSize) {
                        throw SockudoException.DeltaFailure("copy exceeds output file size")
                    }
                    if (offset + count > base.size) {
                        throw SockudoException.DeltaFailure("copy extends past end of input")
                    }
                    repeat(count) { index -> output.add(base[offset + index]) }
                }

                ':' -> {
                    total += count
                    if (total > outputSize) {
                        throw SockudoException.DeltaFailure("insert command gives an output larger than predicted")
                    }
                    if (reader.position + count > delta.size) {
                        throw SockudoException.DeltaFailure("insert count exceeds size of delta")
                    }
                    repeat(count) { output.add(delta[reader.position + it]) }
                    reader.position += count
                }

                ';' -> {
                    val bytes = output.toByteArray()
                    if (count != checksum(bytes)) {
                        throw SockudoException.DeltaFailure("bad checksum")
                    }
                    if (total != outputSize) {
                        throw SockudoException.DeltaFailure("generated size does not match predicted size")
                    }
                    return bytes
                }

                else -> throw SockudoException.DeltaFailure("unknown delta operator")
            }
        }

        throw SockudoException.DeltaFailure("unterminated delta")
    }

    private fun checksum(bytes: ByteArray): Int {
        var sum0 = 0
        var sum1 = 0
        var sum2 = 0
        var sum3 = 0
        var index = 0
        var remaining = bytes.size

        while (remaining >= nHash) {
            sum0 += bytes[index + 0].u() + bytes[index + 4].u() + bytes[index + 8].u() + bytes[index + 12].u()
            sum1 += bytes[index + 1].u() + bytes[index + 5].u() + bytes[index + 9].u() + bytes[index + 13].u()
            sum2 += bytes[index + 2].u() + bytes[index + 6].u() + bytes[index + 10].u() + bytes[index + 14].u()
            sum3 += bytes[index + 3].u() + bytes[index + 7].u() + bytes[index + 11].u() + bytes[index + 15].u()
            index += nHash
            remaining -= nHash
        }

        while (remaining >= 4) {
            sum0 += bytes[index + 0].u()
            sum1 += bytes[index + 1].u()
            sum2 += bytes[index + 2].u()
            sum3 += bytes[index + 3].u()
            index += 4
            remaining -= 4
        }

        sum3 += (sum2 shl 8) + (sum1 shl 16) + (sum0 shl 24)
        when (remaining) {
            3 -> {
                sum3 += bytes[index + 2].u() shl 8
                sum3 += bytes[index + 1].u() shl 16
                sum3 += bytes[index + 0].u() shl 24
            }

            2 -> {
                sum3 += bytes[index + 1].u() shl 16
                sum3 += bytes[index + 0].u() shl 24
            }

            1 -> sum3 += bytes[index + 0].u() shl 24
        }

        return sum3
    }

    private fun Byte.u(): Int = toInt() and 0xff
}
